File: android/siopkit/src/main/kotlin/jp/co/pendako/siop/SelfIssuedIdToken.kt
package jp.co.pendako.siop

import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonObject

/** Self-Issued ID Token issuance (OpenID Connect Core 1.0 Section 7.4). */
object SelfIssuedIdToken {
    /** Section 7.4: the issuer of a self-issued ID Token. */
    const val ISSUER = "https://self-issued.me"

    /** How long a token is valid for, as the consent screen tells the user. */
    const val LIFETIME_SECONDS = 600L

    fun issue(
        request: AuthorizationRequest,
        key: SiopKeyProvider,
        expiresInSeconds: Long = LIFETIME_SECONDS,
        nowEpochSeconds: Long = System.currentTimeMillis() / 1000,
    ): String {
        val jwk = key.publicJwk()
        val claims = buildJsonObject {
            put("iss", ISSUER)
            put("sub", jwk.thumbprint())
            putJsonObject("sub_jwk") {
                jwk.toJsonObject().forEach { (name, value) -> put(name, value) }
            }
            put("aud", request.clientId)
            put("nonce", request.nonce)
            put("iat", nowEpochSeconds)
            put("exp", nowEpochSeconds + expiresInSeconds)
        }
        return Jws.signRs256(claims, key)
    }
}

File: ios/SIOPKit/Sources/SIOPKit/SelfIssuedIDToken.swift
import Foundation

/// Self-Issued ID Token issuance (OpenID Connect Core 1.0 Section 7.4).
public enum SelfIssuedIDToken {
    /// Section 7.4: the issuer of a self-issued ID Token.
    public static let issuer = "https://self-issued.me"

    /// How long an issued token stays valid, unless the caller says otherwise.
    public static let lifetime: TimeInterval = 600

    public static func issue(
        for request: AuthorizationRequest,
        key: SIOPKeyProvider,
        expiresIn: TimeInterval = lifetime,
        now: Date = Date()
    ) throws -> String {
        let jwk = try key.publicJWK()
        let claims: [String: Any] = [
            "iss": issuer,
            "sub": jwk.thumbprint(),
            "sub_jwk": jwk.jsonObject,
            "aud": request.clientID,
            "nonce": request.nonce,
            "iat": Int(now.timeIntervalSince1970),
            "exp": Int(now.addingTimeInterval(expiresIn).timeIntervalSince1970),
        ]
        return try JWS.signRS256(payload: claims, key: key)
    }
}
